lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -127,7 +127,33 @@ def inline_metadata(catalog, checksum_type)
       file = resource.to_ral
 
       if file.recurse?
-        # TODO: find and replace metadata for children recursively
+        file.recurse_remote_metadata do |meta|
+          # Don't create a new resource for the parent directory
+          next if meta.relative_path == "."
+
+          child_resource = Puppet::Resource.new(:file, File.join(file[:path], meta.relative_path))
+          child_resource[:ensure] = meta.ftype
+          # TODO: Make the coying of owner, group, and mode conditional on using permissions
+          child_resource[:owner]  = meta.owner.to_s
+          child_resource[:group]  = meta.group.to_s
+          child_resource[:mode]   = meta.mode.to_s(8).rjust(4, '0')
+          child_resource[:source] = meta.source
+
+          if meta.ftype == "file"
+            child_resource[:checksum_value] = sumdata(meta.checksum)
+            child_resource[:checksum] = meta.checksum_type
+          end
+
+          # Copy parameters from original parent directory
+          file.original_parameters.each do |param, value|
+            # These should never be passed to our children
+            unless [:parent, :ensure, :recurse, :recurselimit, :target, :alias, :source].include? param
+              child_resource[param] = value.to_s
+            end
+          end
+          catalog.add_resource(child_resource)
+          catalog.add_edge(file, child_resource)
```

Comment:
I think it's fine to leave this part as is. We can revisit the "create an edge between the file and its nearest ancestor" and "creating edges from the generated resource to resources that depend on the original directory"